### PR TITLE
Add guidelines for api design

### PR DESCRIPTION
**Pull request by Steven-Chan**
connect #207

### CODE REVIEW

File: features/207-sdk-api-consistency/record-api.md [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+typedef void (^SKYRecordSaveCompletion)(SKYRecord *record, NSError *error);
+
+- (void)saveRecord:(SKYRecord *)record
+        completion:(SKYRecordSaveCompletion)completion;
+
+- (void)saveRecords:(NSArray<SKYRecord *> *)records
+        completionHandler:(void (^)(NSArray *savedRecords,
+                                    NSError *operationError))completionHandler;
+
+- (void)saveRecordsInatomically:(NSArray<SKYRecord *> *)records
```
Comment (carmenlau) [inline features/207-sdk-api-consistency/record-api.md]: Should we use nonatomic?

File: features/207-sdk-api-consistency/record-api.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+- (void)deleteRecordsWithIDsAtomically:(NSArray<SKYRecordID *> *)recordIDs
+                     completionHandler:
+                         (void (^_Nullable)(NSArray *_Nullable deletedRecordIDs,
+                                            NSError *_Nullable error))completionHandler;
+```
+
+##### New
+
+```objc
+- (void)deleteRecordWithID:(SKYRecordID *)recordID
```
Comment (carmenlau) [inline features/207-sdk-api-consistency/record-api.md]: I think all the delete api, we need `recordType`?

**Comment by Steven-Chan** [comment]
@carmenlau updated

File: features/207-sdk-api-consistency/guidelines.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+iOS: `[container pubsub].delegate = xxx`
+Android: `container.getPubsub().setListener(xxx)`
+
+##### Single callback function
+
+iOS: `[[container pubsub] subscribeTo:@"xxx" handler:xxx]`
+Android: `container.getPubsub().subscribe("xxx", xxx)`
+
+While in JS, `skygear.pubsub.on('xxx', function() { xxx })`.
```
[the remark this answers is not in the document]
Comment (chpapa) [reply features/207-sdk-api-consistency/guidelines.md]: 👍

File: features/207-sdk-api-consistency/guidelines.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+# Exceptional cases
+
+### Query in iOS
+
+iOS (ObjC/Swift) provides predicate class(es) natively.
+
+### Record save, mutliple or singular
+
+`save(Records)` and `save(Record)` should be separate functions.
```
[the remark this answers is not in the document]
Comment (chpapa) [reply features/207-sdk-api-consistency/guidelines.md]: `saveSingle` sounds very werid...

[the remark this answers is not in the document]
Comment (carmenlau) [reply features/207-sdk-api-consistency/guidelines.md]: I think we can remove this part, as steven added detailed guideline for record save already. Please read `record-api.md` in this PR and post comment there. The record api looks good for me. 

@Steven-Chan I think this part is not valid now, please help remove it :)

File: features/207-sdk-api-consistency/guidelines.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+For API with same purpose / calling the same skygear-server API, the same set of vocabulary should be used.
+
+<table>
+  <tr>
+    <td>iOS</td>
+    <td>Android</td>
+    <td>JS</td>
+  </tr>
+  <tr>
+    <td>[container signupAnonymouslyWithCompletionHandler:]</td>
```
[the remark this answers is not in the document]
Comment (Steven-Chan) [reply features/207-sdk-api-consistency/guidelines.md]: should be `completion:` instead?

**Comment by Steven-Chan** [comment]
@cheungpat @carmenlau updated

**Pull request merged**